### PR TITLE
[WIP] Introduce kubeapiserver.config.k8s.io/v1 and use standard method for parsing encryption config file

### LINKED ISSUE #61599: EncryptionConfig: use standard method for parsing config file
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature
/sig auth

**What happened**:
Currently encryption config parsing uses a non-standard parsing method that pays no attention to the version of the config:

https://github.com/kubernetes/kubernetes/blob/dce1b881284a103909f5cfa969ff56e5e0565362/staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go#L67-L79

**What you expected to happen**:
- the types must change to use standard methods for including kind/apiVersion (inlining metav1.TypeMeta like all other objects)
- define a versioned config (we can promote straight to v1beta1 if we don't require any structural changes)
- parsing must require well-formed config files

**Anything else we need to know?**:
Requested by @liggitt and extracted from https://github.com/kubernetes/kubernetes/pull/61592#issuecomment-375655819

Steps to be done:
- [ ] fix type definition to actually be a runtime object, inline TypeMeta, etc
- [ ] define v1beta1 API types, generate conversions/defaults
- [ ] define scheme/codecs that have the external (v1beta1) and internal (existing) types registered
- [ ] use that scheme/codec to load the config (handles decoding the v1beta1 version, conversion to internal version)

CC @marrrvin

### PR DESCRIPTION
**Pull request by php-coder**
**What this PR does / why we need it**:
This PR reworks handling of the configuration file for encryption at rest. Now it uses a standard approach for parsing and also it supports versioning.

**Which issue(s) this PR fixes**:
Fixes #61599

### CODE REVIEW

**Comment by php-coder** [comment]
/release-note-none
CC @simo5 @marrrvin

**Comment by php-coder** [comment]
/sig auth

At this time it doesn't work yet as compilation fails with:
```
import cycle not allowed
package k8s.io/kubernetes/cmd/hyperkube
	imports k8s.io/kubernetes/cmd/kube-apiserver/app
	imports k8s.io/apiserver/pkg/server/options/encryptionconfig
	imports k8s.io/apiserver/pkg/server/options/encryptionconfig/scheme
	imports k8s.io/apiserver/pkg/server/options/encryptionconfig
```

@liggitt @smarterclayton We need to move encryption configuration somewhere, but I'm not sure where its place. If we want to use `apiserver.config.k8s.io` API Group then we should move encryption config to `https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiserver/pkg/apis/apiserver` but it's in `v1alpha1` version. I'm not sure that we want and can downgrade api version from v1 to v1alpha1.

How we can handle this?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go [hunk 1]
```diff
@@ -77,10 +79,10 @@ type IdentityConfig struct{}
 // KMSConfig contains the name, cache size and path to configuration file for a KMS based envelope transformer.
 type KMSConfig struct {
 	// name is the name of the KMS plugin to be used.
-	Name string `json:"name"`
+	Name string
 	// cacheSize is the maximum number of secrets which are cached in memory. The default value is 1000.
 	// +optional
-	CacheSize int `json:"cachesize,omitempty"`
```
Comment (php-coder) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: @liggitt This type looks suspicious. Did we decide some time ago to always use integers with a fixed size in APIs, do we?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 2]
```diff
@@ -181,12 +181,12 @@ func GetPrefixTransformers(config *ResourceConfig) ([]value.PrefixTransformer, e
 	return result, nil
 }
 
-// BlockTransformerFunc takes an AES cipher block and returns a value transformer.
-type BlockTransformerFunc func(cipher.Block) value.Transformer
+// blockTransformerFunc takes an AES cipher block and returns a value transformer.
+type blockTransformerFunc func(cipher.Block) value.Transformer
```
Comment (php-coder) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: I decided to make some functions private, it doesn't relate to my change. Let me know if it's inappropriate change at all or if I should submit it separately.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 3]
```diff
@@ -44,30 +41,9 @@ const (
...
-// ParseEncryptionConfiguration parses configuration data and returns the transformer overrides
-func ParseEncryptionConfiguration(f io.Reader) (map[schema.GroupResource]value.Transformer, error) {
-	configFileContents, err := ioutil.ReadAll(f)
-	if err != nil {
-		return nil, fmt.Errorf("could not read contents: %v", err)
-	}
-
-	var config EncryptionConfig
-	err = yaml.Unmarshal(configFileContents, &config)
+	config, err := loadConfigFromFile(filepath)
```
Comment (php-coder) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: @liggitt Do we have a requirement to fully support old format of the config file? Perhaps we should show a warning in this case?

**Comment by liggitt** [comment]
> I'm not sure that we want and can downgrade api version from v1 to v1alpha1.

The documentation says to use `v1`:

https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#understanding-the-encryption-at-rest-configuration

Because we were using a non-standard load that ignored version, this wasn't caught before.

I'd actually be in favor of promoting the config version to `v1`. The types are minimal, there's been no desired changes since it was created, and anyone following our documentation is already using `v1`.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go]: the field in the external type should be a fixed size int, yes

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go (hunk 2 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: there's no real reason to make these private, so let's revert this change.

**Comment by php-coder** [comment]
>> I'd actually be in favor of promoting the config version to v1.
>
> I'll be able to prepare and submit PR next week.

Done: https://github.com/kubernetes/kubernetes/pull/63304

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go (hunk 3 above)
Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: I assume that no, we don't have such requirement.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install/install.go [hunk 4]
```diff
@@ -31,11 +32,12 @@ func Install(registry *registered.APIRegistrationManager, scheme *runtime.Scheme
 	if err := announced.NewGroupMetaFactory(
 		&announced.GroupMetaFactoryArgs{
 			GroupName:                  webhookadmission.GroupName,
-			VersionPreferenceOrder:     []string{v1alpha1.SchemeGroupVersion.Version},
+			VersionPreferenceOrder:     []string{v1.SchemeGroupVersion.Version, v1alpha1.SchemeGroupVersion.Version},
```
Comment (php-coder) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install/install.go]: I'm not sure about this change because v1 has one type and v1alpha1 has others, in other words these aren't different version of the same API but different.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 5]
```diff
@@ -100,13 +93,32 @@ func ParseEncryptionConfiguration(f io.Reader) (map[schema.GroupResource]value.T
...
+}
+
+// loadConfig decodes data as a EncryptionConfiguration object.
+func loadConfig(data []byte) (*cfg.EncryptionConfiguration, error) {
+	scheme := runtime.NewScheme()
+	codecs := serializer.NewCodecFactory(scheme)
+	cfg.AddToScheme(scheme)
+	cfgv1.AddToScheme(scheme)
+
+	configObj, gvk, err := codecs.UniversalDecoder().Decode(data, nil, nil)
```
Comment (php-coder) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: Do we have a better way for parsing YAML? I couldn't find a way of getting a scheme.

CC @sttts

File: cluster/gce/config-default.sh [hunk 6]
```diff
@@ -197,8 +197,8 @@ fi
 ENCRYPTION_PROVIDER_CONFIG="${ENCRYPTION_PROVIDER_CONFIG:-}"
 if [[ -z "${ENCRYPTION_PROVIDER_CONFIG}" ]]; then
     ENCRYPTION_PROVIDER_CONFIG=$(cat << EOM | base64 | tr -d '\r\n'
-kind: EncryptionConfig
-apiVersion: v1
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
```
Comment (php-coder) [inline cluster/gce/config-default.sh]: This PR makes the old configs unusable. If we have to be backward compatible and you know how to handle such cases, please, share this knowledge with me or, at least, point me to the right direction. Thanks!

**Comment by php-coder** [comment]
@liggitt PTAL
@deads2k @sttts Your feedback would be welcomed!

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go (hunk 5 above)
Comment (ericchiang) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go]: This is what we do for the audit API policy file https://github.com/kubernetes/kubernetes/blob/484f62a5686f718429907c2fec8406749e2f6260/staging/src/k8s.io/apiserver/pkg/audit/policy/reader.go#L47

File: cluster/gce/config-default.sh (hunk 6 above)
Comment (ericchiang) [reply cluster/gce/config-default.sh]: Definitely needs a release note.

We could try to parse out the kind and apiVersion beforehand, then conditionally use the old parsing logic. Don't know how much work that is.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install/install.go (hunk 4 above)
Comment (ericchiang) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install/install.go]: The admissionregistration seems to do something similar where different versions of the API hold different object.

https://github.com/kubernetes/api/blob/kubernetes-1.10.2/admissionregistration/v1alpha1/register.go#L45-L48
https://github.com/kubernetes/api/blob/kubernetes-1.10.2/admissionregistration/v1beta1/register.go#L45-L50

Any apimachinery folk care to comment here? It looks fine to me.

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -1688,8 +1688,8 @@ function start-kube-apiserver {
 
   if [[ -n "${ETCD_KMS_KEY_ID:-}" ]]; then
     ENCRYPTION_PROVIDER_CONFIG=$(cat << EOM | base64 | tr -d '\r\n'
-kind: EncryptionConfig
-apiVersion: v1
+kind: EncryptionConfiguration
+apiVersion: apiserver.config.k8s.io/v1
```
Comment (sttts) [inline cluster/gce/gci/configure-helper.sh]: is this api group in line with other componentconfigs? It's not really a componentconfig, or is it?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go [hunk 8]
```diff
@@ -27,3 +27,71 @@ type WebhookAdmission struct {
 	// KubeConfigFile is the path to the kubeconfig file.
 	KubeConfigFile string
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EncryptionConfiguration stores the complete configuration for encryption providers.
+type EncryptionConfiguration struct {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: seeing these in webhookadmission/types.go surprises me?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission
+// +k8s:defaulter-gen=TypeMeta
+
+// Package v1 is the v1 version of the API.
+// +groupName=apiserver.config.k8s.io
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: why in this package? Why in the admission plugin? Is this for admission or general apiserver?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go [hunk 10]
```diff
@@ -14,50 +14,51 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package encryptionconfig
+package v1
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go]: why is this moved?

File: cluster/gce/gci/configure-helper.sh (hunk 7 above)
Comment (php-coder) [reply cluster/gce/gci/configure-helper.sh]: @sttts I looked at the `kubelet.config.k8s.io` API group when I made this PR. How to ensure that this PR "in line with other componentconfigs"?

> It's not really a componentconfig, or is it?

Yes, it's need to set up API server for encrypting/decrypting resources when they are being read/write from/to database.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go (hunk 8 above)
Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: Yeah, I share your feeling. I've just put it near the `v1alpha1` group. Where it should be?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go (hunk 9 above)
Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: @sttts 

> why in this package?

We had discussion with @liggitt about where to place EncryptionConfiguration and concluded that this is the best place. See: https://github.com/kubernetes/kubernetes/issues/61599#issuecomment-375711161

> Why in the admission plugin?

The only reason is because `apiserver.config.k8s.io` was here from its inception.

> Is this for admission or general apiserver?

It doesn't relate to admission at all. I'm not sure what is the "general apiserver" but it definitely related to api server.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go (hunk 10 above)
Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go]: Because the old type was removed as no one uses it anymore.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go (hunk 8 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/types.go]: It's own API group. This has nothing todo with the webhooks.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: > The only reason is because apiserver.config.k8s.io was here from its inception.

We should move it to a more top-level place, e.g. pkg/apis/config.

File: cluster/gce/gci/configure-helper.sh (hunk 7 above)
Comment (sttts) [reply cluster/gce/gci/configure-helper.sh]: That was my question, so `apiserver.config.k8s.io` is in line with it. Interestingly, the scheduler uses `componentconfig/v1alpha1`. That group is a big mess.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go (hunk 10 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go]: And is this v1 quality? I doubt that. More like v1alpha1 with the many changes made.

Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/types.go]: We have to stick with v1 because it's already documented and users already use it: https://github.com/kubernetes/kubernetes/pull/63032#issuecomment-383705778

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go (hunk 9 above)
Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: After I moved it to `pkg/apis/componentconfig`, it started failing with:
```
2018/05/10 14:38:48 Inspecting imports under ./vendor/k8s.io/apiserver/...
2018/05/10 14:38:49 - validating imports for 129 packages in the tree
2018/05/10 14:38:49 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission:
2018/05/10 14:38:49 --- k8s.io/kubernetes/pkg/apis/componentconfig
2018/05/10 14:38:49 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install:
2018/05/10 14:38:49 --- k8s.io/kubernetes/pkg/apis/componentconfig/v1
2018/05/10 14:38:49 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig:
2018/05/10 14:38:49 --- k8s.io/kubernetes/pkg/apis/componentconfig
2018/05/10 14:38:49 --- k8s.io/kubernetes/pkg/apis/componentconfig/v1
2018/05/10 14:38:49 - FAIL
```

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: apiserver config logically belongs in k8s.io/apiserver. by "top level", I think @sttts meant something like `k8s.io/apiserver/pkg/apis`, etc

Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: @liggitt Thanks, Jordan! It looks like that `k8s.io/apiserver/pkg/apis` contains files for publicly available APIs that have REST endpoints and can be consumed by end users. If it's true, then I don't see a reason for putting component' configuration there.

Also I see that components like `KubeSchedulerConfiguration`, `CloudProviderConfiguration`, `VolumeConfiguration`, etc aren't presented there at all.

So, how the things should be, to make this PR mergeable?

Comment (php-coder) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: @sttts What will be my next step then? Is it "move internal and v1 types of apiserver.config.k8s.io to `vendor/k8s.io/apiserver/pkg/apis/componentconfig`"? (BTW what examples are doing there?!)

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/v1/doc.go]: Yes, move them. Then replace the loading code with a proper codec to internal types and then add a test that both old and new api groups can be loaded.

File: pkg/apis/componentconfig/types.go [hunk 11]
```diff
@@ -554,3 +554,71 @@ const (
 	// "kube-scheduler" is the default scheduler lock object name
 	SchedulerDefaultLockObjectName = "kube-scheduler"
 )
+
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: others are moving out of here. We need a plan where componentconfigs should go.

I still think something like k8s.io/componentconfig/kube-apiserver for the versioned types and cmd/kube-apiserver/.... for internal types makes most sense.

/cc @mtaufen @luxas @liggitt

File: pkg/apis/componentconfig/types.go [hunk 12]
```diff
@@ -554,3 +554,71 @@ const (
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// EncryptionConfiguration stores the complete configuration for encryption providers.
+type EncryptionConfiguration struct {
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: this name is far too generic for a global API group.

File: pkg/apis/componentconfig/v1/register.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+package v1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name use in this package.
+const GroupName = "apiserver.config.k8s.io"
```
Comment (sttts) [inline pkg/apis/componentconfig/v1/register.go]: this is a generic componentconfig API group here, not specific to apiservers.

File: pkg/apis/componentconfig/types.go (hunk 11 above)
Comment (luxas) [reply pkg/apis/componentconfig/types.go]: Why `cmd/kube-apiserver/...`, not `pkg/kubeapiserver/apis/apiserverconfig` like the kubelet for example?
`pkg` must not depend on `cmd` and `pkg` should depend on internal types, right?
`k8s.io/componentconfig/...` for the external types SGTM

File: cluster/gce/gci/configure-helper.sh [hunk 14]
```diff
@@ -2609,4 +2609,4 @@ if [[ "$#" -eq 1 && "${1}" == "--source-only" ]]; then
    :
 else
    main "${@}"
-fi
+fi
```
Comment (luxas) [inline cluster/gce/gci/configure-helper.sh]: I don't think you meant to change this

File: pkg/apis/componentconfig/v1/register.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (luxas) [inline pkg/apis/componentconfig/v1/register.go]: New files should be marked 2018

File: pkg/apis/componentconfig/v1/doc.go [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission
+// +k8s:defaulter-gen=TypeMeta
+
+// Package v1 is the v1 version of the API.
+// +groupName=apiserver.config.k8s.io
+package v1
```
Comment (luxas) [inline pkg/apis/componentconfig/v1/doc.go]: This is in the wrong package, as @sttts pointed out. No more API types should be added here, and this is a catch-all directory now which we're moving away from. I think `pkg/kubeapiserver/apis/kubeapiserverconfig` makes sense in case it's kube-apiserver-specific or `k8s.io/apiserver/apis/apiserverconfig` if it's generic for apiservers.

**Review by luxas — CHANGES_REQUESTED** [review]
Need to agree on the place for this first.
No more types must be added in this legacy componentconfig directory.

File: pkg/apis/componentconfig/types.go (hunk 11 above)
Comment (mtaufen) [reply pkg/apis/componentconfig/types.go]: since the API groups are `*.config.k8s.io`, perhaps we should use `k8s.io/config/*` as the path convention when we publish to a staging repo?

File: pkg/apis/componentconfig/v1/doc.go (hunk 16 above)
Comment (php-coder) [reply pkg/apis/componentconfig/v1/doc.go]: > I think `pkg/kubeapiserver/apis/kubeapiserverconfig` makes sense in case it's kube-apiserver-specific

`pkg/kubeapiserver/apis/kubeapiserverconfig` doesn't exist, should I create it then?

> or `k8s.io/apiserver/apis/apiserverconfig` if it's generic for apiservers

`k8s.io/apiserver/apis/apiserverconfig` doesn't exist either. Did you mean `vendor/k8s.io/apiserver/pkg/apis/apiserverconfig`? In this case, `apiserverconfig` sub directory also should be created.

File: cluster/gce/gci/configure-helper.sh (hunk 14 above)
Comment (php-coder) [reply cluster/gce/gci/configure-helper.sh]: Yes, and I've reverted this change already. Perhaps, you were reviewing commits instead of a full changes.

File: pkg/apis/componentconfig/types.go (hunk 12 above)
Comment (php-coder) [reply pkg/apis/componentconfig/types.go]: > this name is far too generic for a global API group.

What did you mean by "a global API group"? It's not global as far I understand.
Do you have a better name instead?

File: pkg/apis/componentconfig/v1/register.go (hunk 13 above)
Comment (php-coder) [reply pkg/apis/componentconfig/v1/register.go]: Ok, I'm going to use `kubeapiserver.config.k8s.io` then.

**Comment by php-coder** [comment]
Hm, compilation fails with missing DeepCopyObject method:
```
W0604 14:06:49.441] + make generated_files
I0604 14:07:13.477] +++ [0604 14:07:13] Building go targets for linux/amd64:
I0604 14:07:13.478]     ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
I0604 14:07:22.934] +++ [0604 14:07:22] Building go targets for linux/amd64:
I0604 14:07:22.934]     ./vendor/k8s.io/code-generator/cmd/defaulter-gen
I0604 14:07:29.356] +++ [0604 14:07:29] Building go targets for linux/amd64:
I0604 14:07:29.356]     ./vendor/k8s.io/code-generator/cmd/conversion-gen
I0604 14:07:36.705] +++ [0604 14:07:36] Building go targets for linux/amd64:
I0604 14:07:36.706]     ./vendor/k8s.io/code-generator/cmd/openapi-gen
I0604 14:07:45.137] +++ [0604 14:07:45] Building go targets for linux/amd64:
I0604 14:07:45.138]     ./vendor/github.com/jteeuwen/go-bindata/go-bindata
W0604 14:07:46.403] + go install ./cmd/...
W0604 14:07:59.323] # k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission
W0604 14:07:59.325] vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/register.go:49:36: cannot use kubeapiserverconfig.EncryptionConfiguration literal (type *kubeapiserverconfig.EncryptionConfiguration) as type runtime.Object in argument to scheme.AddKnownTypes:
W0604 14:07:59.326] 	*kubeapiserverconfig.EncryptionConfiguration does not implement runtime.Object (missing DeepCopyObject method)
W0604 14:08:42.358] # k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig
W0604 14:08:42.364] vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:103:2: undefined: kubeapiserverconfig.AddToScheme
W0604 14:08:42.365] vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:110:25: impossible type assertion:
W0604 14:08:42.366] 	*kubeapiserverconfig.EncryptionConfiguration does not implement runtime.Object (missing DeepCopyObject method)
```
Looks like, I forgot to run some script or maybe I need to "register" a path with new type somewhere so the scripts will pick it up?

**Comment by php-coder** [comment]
>W0604 14:07:59.325] vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/register.go:49:36: cannot use kubeapiserverconfig.EncryptionConfiguration literal (type *kubeapiserverconfig.EncryptionConfiguration) as type runtime.Object in argument to scheme.AddKnownTypes:
>
>Looks like, I forgot to run some script or maybe I need to "register" a path with new type somewhere so the scripts will pick it up?

I've fixed this by adding `doc.go`, `register.go` and `install/install.go` files for an internal type.

---

At this moment, I faced with another verify import violation:
```
2018/06/15 10:18:27 Inspecting imports under ./vendor/k8s.io/apiserver/...
2018/06/15 10:18:28 - validating imports for 130 packages in the tree
2018/06/15 10:18:28 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission:
2018/06/15 10:18:28 --- k8s.io/kubernetes/pkg/kubeapiserver/apis/kubeapiserverconfig
2018/06/15 10:18:28 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/plugin/webhook/config/apis/webhookadmission/install:
2018/06/15 10:18:28 --- k8s.io/kubernetes/pkg/kubeapiserver/apis/kubeapiserverconfig/v1
2018/06/15 10:18:28 -- found forbidden imports for k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/options/encryptionconfig:
2018/06/15 10:18:28 --- k8s.io/kubernetes/pkg/kubeapiserver/apis/kubeapiserverconfig/v1
2018/06/15 10:18:28 --- k8s.io/kubernetes/pkg/kubeapiserver/apis/kubeapiserverconfig
2018/06/15 10:18:28 - FAIL
```
Not sure how to fix it, perhaps, I need to relax this check.

**Comment by luxas** [comment]
This proposal is probably a bit related to this PR: https://github.com/kubernetes/community/pull/2354

**Comment by stlaz** [comment]
@luxas I finally got to read the design and compare it to this PR.

If I understand the KEP correctly, the `kubeapiserver` package would eventually get its own repository. That action still needs to be done. Therefore the only implication for this PR right now is that I should move the changes for the staging package to `kubernetes/apiserver` repository, am I correct?

**Comment by luxas** [comment]
> kubeapiserver package would eventually get its own repository

Yes, eventually, but that's not a near-term goal for that KEP specifically.

>  Therefore the only implication for this PR right now is that I should move the changes for the staging package to kubernetes/apiserver repository, am I correct?

If this is generic apiserver code I guess it can live in `k8s.io/apiserver` if it's not I'd prefer `k8s.io/kube-apiserver`. The `k8s.io/apiserver/pkg/apis/config{,v1alpha1}` packages are already created FWIW.

**Comment by stlaz** [comment]
This PR originally implemented it as a part of `apiserver` but after series of decisions the `kubeapiserver` package was chosen. I'd stick to that unless there is a specific reason to move it back or to `kube-apiserver` instead.
Also, this is going to be `v1` due to https://github.com/kubernetes/kubernetes/pull/63032#issuecomment-383705778

**Comment by stlaz** [comment]
@luxas I just went through the code again, we could probably really go and contain this change just for `apiserver` since we're not touching neither `kube-apiserver` nor `federation-apiserver` codebase. I suspect `kubeapiserver/apis/kubeapiserverconfig` was chosen simply for the sake of no other better location which now hopefulyl exists with the `k8s.io/apiserver/pkg/apis/config` you're mentioning. It seems it would be a code-wise cleaner change, too.

File: cluster/gce/config-default.sh (hunk 6 above)
Comment (sttts) [reply cluster/gce/config-default.sh]: We can register the types under the legacy core group in the scheme used for decoding. Why is that no option?

Comment (sttts) [reply cluster/gce/config-default.sh]: yes, same types, multiple groups. Technically possible.

Good question about deprecation. Formally we will probably have to stick with this forever.

**Comment by php-coder** [comment]
Superseded by #67383

/close
